In [62]:
library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
library(Matrix)


载入程辑包：'Matrix'


The following object is masked from 'package:S4Vectors':

    expand




In [2]:
root_dir = '.'
pbmc_scrna_data_dir = file.path(root_dir, 'pbmc8k_filtered_gene_bc_matrices', 'filtered_gene_bc_matrices', 'GRCh38')
pbmc_atac_data_dir = file.path(root_dir, 'atac_v1_pbmc_10k_filtered_peak_bc_matrix','filtered_peak_bc_matrix')
atac_fragment_path = file.path(root_dir, "atac_v1_pbmc_10k_fragments.tsv.gz")

In [3]:
pbmc_atac_data <- ReadMtx(mtx=file.path(pbmc_atac_data_dir,'matrix.mtx'), 
                          cells=file.path(pbmc_atac_data_dir, 'barcodes.tsv'),
                          features=file.path(pbmc_atac_data_dir, 'peaks.tsv'),
                          feature.column=1)
pbmc_atac_metadata <- read.table("pbmc_atac_metadata.tsv", sep=',')

In [4]:
pbmc_atac_obj <- CreateSeuratObject(counts = pbmc_atac_data, project = "pbmc10k", assay="ATAC", meta.data=pbmc_atac_metadata)

In [5]:
pbmc_atac_metadata

orig.ident    nCount_ATAC nFeature_ATAC
AAACGAAAGAGCGAAA-1 SeuratProject  1172       1036         
AAACGAAAGAGTTTGA-1 SeuratProject  1358       1179         
AAACGAAAGCGAGCTA-1 SeuratProject  2395       1816         
AAACGAAAGGCTTCGC-1 SeuratProject 18295       8919         
AAACGAAAGTGCTGAG-1 SeuratProject   869        781         
AAACGAAGTCAGGCTC-1 SeuratProject   523        456         
AAACGAAGTGCCCGAT-1 SeuratProject  1759       1517         
AAACGAAGTTGTATCG-1 SeuratProject  1248       1106         
AAACGAATCAGTTGAC-1 SeuratProject   802        708         
AAACGAATCCTTACGC-1 SeuratProject  1042        888         
AAACGAATCGTTTGAG-1 SeuratProject  1066        885         
AAACGAATCTGCCGAG-1 SeuratProject   228        219         
AAACTCGAGATGCGCA-1 SeuratProject  2051       1673         
AAACTCGAGCATGATA-1 SeuratProject   712        653         
AAACTCGAGCGCATTT-1 SeuratProject  1936       1469         
AAACTCGAGTACTCTG-1 SeuratProject  1859       1315         
AAACTCGAGTCACGCC-1 SeuratProject  1023        966         
AAACTCGAGTCTCTAG-1 SeuratProject  1365       1184         
AAACTCGCAAAGAGAG-1 SeuratProject  3025       2188         
AAACTCGCAAGCGAAC-1 SeuratProject   923        785         
AAACTCGCACTCAGCG-1 SeuratProject  1133        993         
AAACTCGCACTTGCGG-1 SeuratProject   698        634         
AAACTCGCAGCGTCGT-1 SeuratProject  1042        833         
AAACTCGCATCACAGT-1 SeuratProject  1108        966         
AAACTCGCATGTGGGA-1 SeuratProject  7100       3993         
AAACTCGGTCAGAAGC-1 SeuratProject  1084        948         
AAACTCGGTCATAGCT-1 SeuratProject  1599       1387         
AAACTCGGTGGATTCT-1 SeuratProject  1248       1065         
AAACTCGGTTTGATCG-1 SeuratProject  3883       2681         
AAACTCGTCAGGTCTA-1 SeuratProject  3898       2772         
⋮                  ⋮             ⋮           ⋮            
TTTGGTTGTACAACGG-1 SeuratProject   812        729         
TTTGGTTGTCATAAGA-1 SeuratProject  1760       1373         
TTTGGTTGTGAATCAA-1 SeuratProject  1249        974         
TTTGGTTGTGCCCGAT-1 SeuratProject   447        395         
TTTGGTTGTTAGGCTT-1 SeuratProject   297        274         
TTTGGTTGTTCTTATC-1 SeuratProject   237        226         
TTTGGTTGTTGCCGCA-1 SeuratProject   127        123         
TTTGGTTTCCAATAGC-1 SeuratProject   746        664         
TTTGGTTTCCTGTGGG-1 SeuratProject  2276       1763         
TTTGGTTTCGACCCGA-1 SeuratProject  1276       1106         
TTTGGTTTCGACTTAG-1 SeuratProject 10755       5259         
TTTGGTTTCGAGTTAC-1 SeuratProject   789        688         
TTTGGTTTCTCGTAGA-1 SeuratProject   547        481         
TTTGTGTAGCATACCT-1 SeuratProject  1411       1223         
TTTGTGTAGCTGGAGT-1 SeuratProject   909        775         
TTTGTGTAGGATGCCA-1 SeuratProject   461        416         
TTTGTGTAGTCCAGAG-1 SeuratProject  2093       1593         
TTTGTGTCAACCTCCT-1 SeuratProject   757        665         
TTTGTGTCAATGCATC-1 SeuratProject  1393       1150         
TTTGTGTCACGGCCAT-1 SeuratProject   723        626         
TTTGTGTGTAACCCAT-1 SeuratProject   573        505         
TTTGTGTGTCGGCTGT-1 SeuratProject   946        828         
TTTGTGTGTGATGCTT-1 SeuratProject   941        821         
TTTGTGTGTTTAAGGA-1 SeuratProject  1306       1134         
TTTGTGTTCATCGCCT-1 SeuratProject   939        851         
TTTGTGTTCATGCTTT-1 SeuratProject  1013        893         
TTTGTGTTCCGAGAGA-1 SeuratProject  1035        902         
TTTGTGTTCGAAGCCC-1 SeuratProject  1528       1306         
TTTGTGTTCTACTTTG-1 SeuratProject  2195       1654         
TTTGTGTTCTTGTGCC-1 SeuratProject  1346       1155         
                   predicted.celltype.l1.score predicted.celltype.l1
AAACGAAAGAGCGAAA-1 0.3726037                   Mono                 
AAACGAAAGAGTTTGA-1 0.6990826                   Mono                 
AAACGAAAGCGAGCTA-1 0.3192837                   Mono                 
AAACGAAAGGCTTCGC-1 0.4638083                   Mono                 
AAACGAAAGTGCTGAG

In [6]:
print(unique(pbmc_atac_obj@meta.data$predicted.celltype.l1))

[1] "Mono"    "DC"      "B"       "CD4 T"   "CD8 T"   "other T" "NK"     
[8] "other"  


In [7]:
print(unique(pbmc_atac_obj@meta.data$predicted.celltype.l2))

 [1] "CD14 Mono"      "pDC"            "B naive"        "NK"            
 [5] "B intermediate" "CD4 Naive"      "CD8 Naive"      "MAIT"          
 [9] "CD4 TCM"        "CD16 Mono"      "Doublet"        "CD8 TEM"       
[13] "cDC2"           "B memory"       "HSPC"           "NK_CD56bright" 
[17] "dnT"            "Platelet"       "Treg"           "Plasmablast"   
[21] "Eryth"          "ASDC"           "CD8 TCM"        "gdT"           
[25] "cDC1"           "CD4 TEM"       


In [8]:
print(unique(pbmc_atac_obj@meta.data$predicted.celltype.l3))

 [1] "CD14 Mono"             "pDC"                   "B naive lambda"       
 [4] "B intermediate kappa"  "CD4 Naive"             "Doublet"              
 [7] "MAIT"                  "CD4 TCM_1"             "NK_4"                 
[10] "CD16 Mono"             "CD4 TCM_2"             "CD8 Naive_2"          
[13] "B naive kappa"         "NK_2"                  "CD8 TEM_2"            
[16] "cDC2_1"                "B memory kappa"        "CD8 TEM_1"            
[19] "HSPC"                  "CD8 Naive"             "NK_CD56bright"        
[22] "dnT_2"                 "Platelet"              "Treg Memory"          
[25] "B intermediate lambda" "NK_1"                  "CD8 TEM_5"            
[28] "CD4 TCM_3"             "CD8 TEM_6"             "B memory lambda"      
[31] "Plasma"                "Eryth"                 "ASDC_pDC"             
[34] "CD8 TEM_4"             "cDC1"                  "CD4 TEM_1"            
[37] "CD4 CTL"               "gdT_1"                


In [9]:
cell_types = unique(pbmc_atac_obj@meta.data$predicted.celltype.l2)
summary(as.factor(pbmc_atac_obj@meta.data$predicted.celltype.l2))

ASDC B intermediate       B memory        B naive      CD14 Mono 
             3            284            125            382           3906 
     CD16 Mono      CD4 Naive        CD4 TCM        CD4 TEM      CD8 Naive 
           211            172            897              1            464 
       CD8 TCM        CD8 TEM           cDC1           cDC2            dnT 
             2            256              2             84            114 
       Doublet          Eryth            gdT           HSPC           MAIT 
           614              3              2            133            104 
            NK  NK_CD56bright            pDC    Plasmablast       Platelet 
           320             83            383              9             39 
          Treg 
            40

* CD14 Mono: 3906
* CD4 TCM: 897
* CD8 Naive: 464
* B naive: 382
* CD8 TEM: 256
* NK: 320

## 筛选peak

In [73]:
save_dir = file.path("pbmc_sep_data")

### CD14 Mono

In [66]:
tgt_type = "CD14 Mono"
sub_data = subset(pbmc_atac_obj, predicted.celltype.l2==tgt_type)
percentile = FindTopFeatures(sub_data@assays$ATAC@counts)

In [41]:
threshold = 0.85
sum(percentile$percentile>threshold)

[1] 12044

In [70]:
cd14_sub_data = subset(sub_data, feature=rownames(subset(percentile, percentile>threshold)))
sp_data = Matrix(cd14_sub_data@assays$ATAC@counts, sparse = TRUE)
cd14_sub_data

An object of class Seurat 
12044 features across 3906 samples within 1 assay 
Active assay: ATAC (12044 features, 0 variable features)
 2 layers present: counts, data

In [76]:
tgt_dir = file.path(save_dir, tgt_type, 'atac')
if (!file.exists(tgt_dir)){
    dir.create(tgt_dir, recursive = TRUE)
}
write(x=sp_data@Dimnames[[1]], file=file.path(tgt_dir, "peaks.tsv"))
write(x=sp_data@Dimnames[[2]], file=file.path(tgt_dir, "barcodes.tsv"))
writeMM(obj=sp_data, file=file.path(tgt_dir, "matrix.mtx"))

NULL

### CD4 TCM

In [77]:
tgt_type = "CD4 TCM"
sub_data = subset(pbmc_atac_obj, predicted.celltype.l2==tgt_type)
percentile = FindTopFeatures(sub_data@assays$ATAC@counts)

In [78]:
threshold = 0.85
sum(percentile$percentile>threshold)

[1] 12040

In [79]:
cd4_sub_data = subset(sub_data, feature=rownames(subset(percentile, percentile>threshold)))
sp_data = Matrix(cd4_sub_data@assays$ATAC@counts, sparse = TRUE)
cd4_sub_data

An object of class Seurat 
12040 features across 897 samples within 1 assay 
Active assay: ATAC (12040 features, 0 variable features)
 2 layers present: counts, data

In [80]:
tgt_dir = file.path(save_dir, tgt_type, 'atac')
if (!file.exists(tgt_dir)){
    dir.create(tgt_dir, recursive = TRUE)
}
write(x=sp_data@Dimnames[[1]], file=file.path(tgt_dir, "peaks.tsv"))
write(x=sp_data@Dimnames[[2]], file=file.path(tgt_dir, "barcodes.tsv"))
writeMM(obj=sp_data, file=file.path(tgt_dir, "matrix.mtx"))

NULL

### CD8 Naive

In [81]:
tgt_type = "CD8 Naive"
sub_data = subset(pbmc_atac_obj, predicted.celltype.l2==tgt_type)
percentile = FindTopFeatures(sub_data@assays$ATAC@counts)

In [82]:
threshold = 0.85
sum(percentile$percentile>threshold)

[1] 12036

In [83]:
cd8n_sub_data = subset(sub_data, feature=rownames(subset(percentile, percentile>threshold)))
sp_data = Matrix(cd8n_sub_data@assays$ATAC@counts, sparse = TRUE)
cd8n_sub_data

An object of class Seurat 
12036 features across 464 samples within 1 assay 
Active assay: ATAC (12036 features, 0 variable features)
 2 layers present: counts, data

In [84]:
tgt_dir = file.path(save_dir, tgt_type, 'atac')
if (!file.exists(tgt_dir)){
    dir.create(tgt_dir, recursive = TRUE)
}
write(x=sp_data@Dimnames[[1]], file=file.path(tgt_dir, "peaks.tsv"))
write(x=sp_data@Dimnames[[2]], file=file.path(tgt_dir, "barcodes.tsv"))
writeMM(obj=sp_data, file=file.path(tgt_dir, "matrix.mtx"))

NULL

### B naive

In [87]:
tgt_type = "B naive"
sub_data = subset(pbmc_atac_obj, predicted.celltype.l2==tgt_type)
percentile = FindTopFeatures(sub_data@assays$ATAC@counts)

In [88]:
threshold = 0.85
sum(percentile$percentile>threshold)

[1] 12079

In [89]:
bn_sub_data = subset(sub_data, feature=rownames(subset(percentile, percentile>threshold)))
sp_data = Matrix(bn_sub_data@assays$ATAC@counts, sparse = TRUE)
bn_sub_data

An object of class Seurat 
12079 features across 382 samples within 1 assay 
Active assay: ATAC (12079 features, 0 variable features)
 2 layers present: counts, data

In [90]:
tgt_dir = file.path(save_dir, tgt_type, 'atac')
if (!file.exists(tgt_dir)){
    dir.create(tgt_dir, recursive = TRUE)
}
write(x=sp_data@Dimnames[[1]], file=file.path(tgt_dir, "peaks.tsv"))
write(x=sp_data@Dimnames[[2]], file=file.path(tgt_dir, "barcodes.tsv"))
writeMM(obj=sp_data, file=file.path(tgt_dir, "matrix.mtx"))

NULL